### ISSUE TITLE
Meterpreter / Tunneling kills the tunnel on high traffic.

### ISSUE THREAD

**Issue by vysec** [rapid7/metasploit-framework#6514]
Hi there,

I don't know if it's a previously reported issue but... here goes.

I have been playing with the PentestIt labs recently and been making use of meterpreter sessions in the following way to pivot and scan locally open services.

Meterpreter session open by uploading a linux/x86/meterpreter/reverse_tcp elf file to the host and calling it through a webshell. This then stages into session 1.

$route add 172.16.0.3 255.255.255.0 1

$use auxiliary/server/socks4a
$run

$proxychains dirb http://172.16.0.3

It scans for a while, but then crashes.

I have noticed this also happen if you send too much of ANY traffic through the tunnel. I noticed this back in the days of doing PWK and OSCP also. Back then, I resorted to just using reverse dynamic tunnels via. SSH.

I am wondering if there are known existing stability issues with meterpreter and tunneling as well as any methods to improve it via. existing options that I can set. Exhaustive nmap scans will also crash the tunnel.

A screenshot may help?

![](http://i.imgur.com/MOcDwY7.png)

Example of running MySQL login bruteforce over the tunnel.

![](http://i.imgur.com/2jLTt4t.png)

**Comment by bcook-r7** [comment]
That's a good recipe at least, will see what we can do.

**Comment by vysec** [comment]
Just want to mention, regardless of setting:

`route add 127.0.0.1 255.255.255.255 SESSION`

What I really meant was that whether or not I was scanning local ports through the tunnel, it would still crash. I also scanned other IPs in the local range such as 172.16.0.3/24.

Thanks for dipping into this :)

**Comment by tdoan-r7** [comment]
We've verified that this is an issue. 

<img width="588" alt="6514-1" src="https://cloud.githubusercontent.com/assets/1195616/14298390/5c45ca3e-fb4a-11e5-89a4-0358823d65d8.png">

<img width="827" alt="6514-2" src="https://cloud.githubusercontent.com/assets/1195616/14298391/5e40c35c-fb4a-11e5-9639-cfe65273e267.png">

Here is the strace:
strace log for linux/x86/meterpreter/reverse_tcp

```select(4, [3], NULL, NULL, {0, 500000}) = 1 (in [3], left {0, 498122})
read(3, "\27\3\1\0 ", 5)                = 5
read(3, "\2160\32`|\307$?4\6\315%\217M\4\372\224Z\345*\22<G\35\304\203@{\16\7\246\337", 32) = 32
read(3, "\27\3\1\1\0", 5)               = 5
read(3, "\3622np\307\202\317F\373\305F\350\32b}\227 \336\315\327\370\374\3321U\35\264(b\364\222\325"..., 256) = 256
mmap2(0x10000000, 1048576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x10000000
mprotect(0x10000000, 4096, PROT_NONE)   = 0
clone(child_stack=0x100fff00, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_DETACHED) = 4804
futex(0x8f1cfb4, FUTEX_WAKE_PRIVATE, 1) = 0
select(4, [3], NULL, NULL, {0, 500000}) = 1 (in [3], left {0, 499524})
read(3, "\25\3\1\0 ", 5)                = 5
read(3, "c\321\2\30/\3010\32v\224O\211gu\337\264\227\267\332Q\316c\23\26o0\375\236\361cru", 32) = 32
futex(0x8f1cbb0, FUTEX_WAKE, 1)         = 0
futex(0x8f1ce40, FUTEX_WAKE, 1)         = 0
futex(0x8f1ce8c, FUTEX_WAIT_PRIVATE, 0, NULL```

@bcook-r7 Maybe you can fix this by adding a recursive mutex?

**Comment by bcook-r7** [comment]
Looks like there is more to do here, reopening.

**Comment by busterb** [comment]
This should be resolved now as of #8279
